store: added option to reencode and compress postings before storing them to the cache

pstibrany opened the pull request:
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This PR introduces ability to compress postings before storing them into cache (in memory or shared cache). It reencodes postings to use more efficient encoding (diff + varint) and then uses snappy to compress it even more. It typically reduces original postings size to about 1/10th. New flag is disabled by default.

## Verification

- unit tests for encoding/decoding postings
- manual testing hitting the code paths in `fetchPostings` method, and stepping through it in debugger

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -1544,7 +1564,20 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
+					data, err := diffVarintSnappyEncode(newBigEndianPostings(pBytes[4:]))
+					if err == nil {
+						storeData = data
+					}
```
What if cache is malformed and will give error?

pstibrany replied:
Here we are encoding postings data fetched from remote storage. If that data is somehow corrupted (`newBigEndianPostings` return error, or postings entries are not in increasing order), we will store original data (as fetched from storage) to the cache. This is what code did already before this PR.

When same corrupted data is later used, queries will return with errors.

bwplotka replied:
Fair :+1:

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,139 @@
+)
+
+const (
+	// these headers should not be prefix of each other
```
Please make sure comments are full sentence (I think `make lint` should detect this)

pstibrany replied:
`make lint` failed on me on my mac :-( But I will check errors by CI.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,139 @@
+	return bytes.HasPrefix(input, []byte(codecHeaderRaw)) || bytes.HasPrefix(input, []byte(codecHeaderSnappy))
+}
+
+func diffVarintSnappyEncode(p index.Postings) ([]byte, error) {
```
Looks quite shallow to me ):

pstibrany replied:
You mean the comment? Well, I can add some, although the function name is pretty self-explanatory to me ;-)

bwplotka replied:
No, I mean function is to shallow. It does not add anything. Abstractions should be deeper and have smaller interfaces otherwise it's just a noise (IMO).

In this case for boolean argument, IDE gives us hint these days so I think we can kill this function.

pstibrany replied:
Ah, I see what you mean. I didn't want to call `diffVarintEncode` with constant `true` argument from bucket.go, so I created this function. You're right that it doesn't add much. False branch (no snappy) is there mostly for testing and to see if compression adds anything or not.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,139 @@
+
+	buf := bytes.Buffer{}
+
+	// if we're returning raw data, write the header to the buffer, and then return buffer directly
```
comment not a full sentence.

bwplotka replied:
and everywhere in this file.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,139 @@
+	return diffVarintEncode(p, true)
+}
+
+func diffVarintEncode(p index.Postings, useSnappy bool) ([]byte, error) {
```
Can we use https://github.com/prometheus/prometheus/blob/cf76daed2f778e97ea70be1bc65ac5f9c5d6d83f/tsdb/encoding/encoding.go#L31 instead of this function?

pstibrany commented:
I've made it through all lint checks 🎉

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -220,6 +220,10 @@ type BucketStore struct {
 	enableIndexHeader        bool
+
+	// Reencode postings using diff+varint+snappy when storing to cache.
+	// This makes them smaller, but takes extra CPU and memory.
```
`memory` is tricky here. Because if you in-mem cache it might reduce overall memory a lot.

pstibrany replied:
Added extra `// When used with in-memory cache, memory usage should decrease overall, thanks to postings being smaller.` sentence to the comment.

bwplotka replied:
Thanks!

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -1472,7 +1483,16 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
-				_, l, err := r.dec.Postings(b)
+				// Even if this instance is not using compression, there may be compressed
+				// entries in the cache written by other stores.
+				var l index.Postings
```
again `var (` available.

pstibrany replied:
OK :) Personally I find `var (` ugly

bwplotka replied:
Curious why, but yea - not a blocker (:

pstibrany replied:
Because of extra indentation. But I've used var block here, as you suggested :)

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -1472,7 +1483,16 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
+				var err error
+				if isDiffVarintEncodedPostings(b) {
+					l, err = diffVarintDecode(b)
+				} else {
```
I would love to avoid `else` everywhere possible... We can do this here as well with duplicated `g.Fill`.. up to you (: It's just whenever I see `else` in Go it's 99% not needed.

pstibrany replied:
> It's just whenever I see else in Go it's 99% not needed.

That's a ... interesting... statement :)

Here we would need to duplicate entire
```
				if err != nil {
					return errors.Wrap(err, "decode postings")
				}
				g.Fill(j, l)
```

block. I feels like extra repetition to me, but whatever works...

bwplotka replied:
> That's a ... interesting... statement :)

Yes, but I can totally prove this (:

bwplotka replied:
Not a blocker here as well, up to you (:

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+)
+
+const (
+	// These headers should not be a prefix of each other.
```
of each other?

bwplotka replied:
what does it mean?

pstibrany replied:
Two headers and comment is now removed.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+
+const (
+	// These headers should not be a prefix of each other.
+	codecHeaderRaw    = "diff+varint+raw"
```
```suggestion
	codecHeaderRaw    = "diff+varint"
```
?

pstibrany replied:
This is exactly what would not work ;-) Since we compare prefix of the input byte array, HasPrefix(input, "diff+varint") returns true for both snappy and not snappy version. I wanted prefix to be unique.

bwplotka replied:
Ok, now is the question why do we need both. Can we only support snappy?

pstibrany replied:
Yes, you're right.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+func diffVarintEncode(p index.Postings, useSnappy bool) ([]byte, error) {
+	buf := encoding.Encbuf{}
+
+	// If we're returning raw data, write the header to the buffer, and then return buffer directly.
```
I don't understand this comment really.. maybe placement is not correct? let's remove it, I totally understand this function without this.

pstibrany replied:
ok

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+	prev := uint64(0)
+	for p.Next() {
+		v := p.At()
+		if v < prev {
```
What if v i `0`?

pstibrany replied:
I *believe* that contract of `index.Postings` is that values are increasing [it is not explicit about it, although the file format is and much of the code using postings requires this property]. That implies that `p.At()` can only be 0 as the very first value, which is ok (`prev` is 0 as well at that point, and `v < prev` is not true).

bwplotka replied:
fair point. Thanks

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+	for p.Next() {
+		v := p.At()
+		if v < prev {
+			// Postings entries must be in increasing order.
```
Not needed comment, it should rather part of error

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+			return nil, errors.Errorf("decreasing entry, val: %d, prev: %d", v, prev)
+		}
+
+		buf.PutUvarint64(v - prev)
```
why we do delta? For better compression?

bwplotka replied:
This one is very much needed to be commented then ;p

pstibrany replied:
Snappy compression on its own doesn't really help, because there is no repetition. Delta makes values small, and Varint is very efficient at encoding small values (values < 128 are encoded as single byte, values < 16384 are encoded as two bytes). It's delta + using varint encoding that reduces size significantly (to about 20% of original), snappy then halves it to 10% of the original.

pstibrany replied:
Well, "diff" (or "delta", perhaps we should rename it) is already a strong indication of what we're doing here :)

bwplotka replied:
:+1: Thanks!

bwplotka replied:
Maybe strong indication but what you said here:

> Snappy compression on its own doesn't really help, because there is no repetition. Delta makes values small, and Varint is very efficient at encoding small values (values < 128 are encoded as single byte, values < 16384 are encoded as two bytes). It's delta + using varint encoding that reduces size significantly (to about 20% of original), snappy then halves it to 10% of the original.

Is important and will be forgotten if it's only on GH PR comment.

pstibrany replied:
Fair enough, I'll put this comment into the file.

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+		buf.PutUvarint64(v - prev)
+		prev = v
+	}
+
```
not needed new line... In the same way we don't do newline between `err := invoke` and `if err != nil`

bwplotka commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ We use *breaking* word for marking changes that are not backward compatible (rel
 ### Added
 
 - [#2265](https://github.com/thanos-io/thanos/pull/2265) Compactor: Add `--wait-interval` to specify compaction wait interval between consecutive compact runs when `--wait` enabled.
+- [#2297](https://github.com/thanos-io/thanos/pull/2297) Store Gateway: Add `--experimental.enable-postings-compression` flag to enable reencoding and compressing postings before storing them into cache. Compressed postings take about 10% of the original size.
```
This is not the only thing that changed. We changed overall the format of cached postings even with flag is set to false, right?

pstibrany replied:
Not really. If flag is set to false, then this store will only write original postings (as found in the index) to the cache. However, if store finds compressed entry in the cache, it will understand it and decompress.

bwplotka replied:
Fair (:

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	raw := input[headerLen:]
+	if compressed {
```
just put it in `case bytes.HasPrefix(input, []byte(codecHeaderSnappy)):`? No need for `compressed` bool

pstibrany replied:
By using bool flag, we can avoid `HasPrefix` comparison which is costly. But I think I'll remove non-snappy version completely, as you suggested in other comment.

bwplotka replied:
yes please (:

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+	return &diffVarintPostings{buf: &encoding.Decbuf{B: raw}}, nil
+}
+
+// Implementation of index.Postings based on diff+varint encoded data.
```
comment should start with struct name

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,134 @@
+		return nil, p.Err()
+	}
+
+	if !useSnappy {
```
hmmm do we really need `useSnappy`? We always use snappy.

pstibrany replied:
It's gone now :)

pracucci commented on pkg/store/bucket.go:
```diff
@@ -1544,7 +1567,20 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
+					// Reencode postings before storing to cache. If that fails, we store original bytes.
+					// This can only fail, if postings data was somehow corrupted,
+					// and there is nothing we can do about it. It's not worth reporting here.
+					data, err := diffVarintSnappyEncode(newBigEndianPostings(pBytes[4:]))
```
Have you considered adding a couple of metrics with postings raw size and the compressed one, to measure the compression ratio in the wild?

bwplotka replied:
Good point

pstibrany replied:
I've added following metrics:
- number of compressions, and compression failures 
- total original size and compressed size
- total compression time

pstibrany replied:
Oh, one more thing... I've moved compression outside of the lock. However, storing data back to cache is still done with lock held, which looks wrong. I haven't fixed that part, because this PR ir not about it, but it looks like it should be moved outside as well. I haven't looked into more details here (like, does it break anything?)

pstibrany replied:
Also added number of decompressions, decompression errors, and total decompression time (snappy part only, we don't time decoding varints and adding values together)

bwplotka commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,135 @@
+// significantly (to about 20% of original), snappy then halves it to ~10% of the original.
+
+const (
+	codecHeaderSnappy = "dvs" // As in "diff+varint+snappy".
```
:+1:

bwplotka commented on cmd/thanos/store.go:
```diff
@@ -81,6 +81,9 @@ func registerStore(m map[string]setupFunc, app *kingpin.Application) {
 	enableIndexHeader := cmd.Flag("experimental.enable-index-header", "If true, Store Gateway will recreate index-header instead of index-cache.json for each block. This will replace index-cache.json permanently once it will be out of experimental stage.").
 		Hidden().Default("false").Bool()
 
+	enablePostingsCompression := cmd.Flag("experimental.enable-postings-compression", "If true, Store Gateway will reencode and compress postings before storing them into cache. Compressed postings take about 10% of the original size.").
```
good point, but since we don't have any other postings storage it's fine for me in smaller form.

bwplotka replied:
Also I assume at some point we will enable it by default and remove this flag

pstibrany replied:
Bartek, if you don't mind, I will change it. It will be clearer.

bwplotka replied:
Sure, go for it :+1:

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -1544,7 +1567,20 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
+					// This can only fail, if postings data was somehow corrupted,
+					// and there is nothing we can do about it. It's not worth reporting here.
+					data, err := diffVarintSnappyEncode(newBigEndianPostings(pBytes[4:]))
+					if err == nil {
```
We will know as querying would fail?

pstibrany replied:
That was my thinking here... if it we get error here, it means postings was corrupted, so it will still be corrupted few lines later when it's used, and error will be reported there.

pstibrany replied:
These errors are now tracked via metric.

pstibrany commented on pkg/store/postings_codec.go:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	// Make result buffer large enough to hold our header and compressed block.
+	result := make([]byte, len(codecHeaderSnappy)+snappy.MaxEncodedLen(len(buf)))
```
No, because both `copy` and `snappy.Encode` only checks for length of the supplied buffer, not its capacity.

pstibrany commented:
> Awesome. LGTM! Will merge after a conflict is resolved 👍 👍 👍

Rebased, and also added metrics for compression and decompression (count, errors, time, original vs compressed size).

bwplotka commented:
Thanks for your hard work @pstibrany ! Sorry for the need for some many rebases ):

pstibrany commented:
No worries, thanks for your help Bartek!
